orbit/pkg/useraction/mdm_migration_darwin.go

```diff
@@ -0,0 +1,286 @@
+//go:build darwin
+
+package useraction
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+type swiftDialogExitCode int
+
+const (
+	primaryBtnExitCode   = 0
+	errorExitCode        = 1
+	secondaryBtnExitCode = 2
+	infoBtnExitCode      = 3
+	timeoutExitCode      = 4
+	userQuitExitCode     = 10
+	unknownExitCode      = 99
+)
+
+var mdmMigrationTemplate = template.Must(template.New("mdmMigrationTemplate").Parse(`
+## Migrate to Fleet
+
+To begin, click "Start." Your default browser will open your My Device page.
+
+{{ if .Aggressive }}You {{ else }} Once you start, you {{ end -}} will see this dialog every 15 minutes until you click "Turn on MDM" and complete the instructions.
+
+\\![Image showing the Fleet UI](https://fleetdm.com/images/permanent/mdm-migration-screenshot-768x180@2x.jpg)
+
+Unsure? Contact {{ .OrgInfo.OrgName }} IT [here]({{ .OrgInfo.ContactURL }}).
+`))
+
+var errorTemplate = template.Must(template.New("").Parse(`
+### Something's gone wrong.
+
+Please contact your IT admin [here]({{ .ContactURL }}).
+`))
+
+// baseDialog implements the basic building blocks to render dialogs using
+// swiftDialog.
+type baseDialog struct {
+	path        string
+	interruptCh chan struct{}
+}
+
+func newBaseDialog(path string) *baseDialog {
+	return &baseDialog{path: path, interruptCh: make(chan struct{})}
+}
+
+func (b *baseDialog) CanRun() bool {
+	if _, err := os.Stat(b.path); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// Exit sends the interrupt signal to try and stop the current swiftDialog
+// instance.
+func (b *baseDialog) Exit() {
+	b.interruptCh <- struct{}{}
+	log.Info().Msg("dialog exit message sent")
+}
+
+// render is a general-purpose render method that receives the flags used to
+// display swiftDialog, and starts an asyncronous routine to display the dialog
+// without blocking.
+//
+// The first returned channel sends the exit code returned by swiftDialog, and
+// the second channel is used to send errors.
+func (b *baseDialog) render(flags ...string) (chan swiftDialogExitCode, chan error) {
+	exitCodeChan := make(chan swiftDialogExitCode, 1)
+	errChan := make(chan error, 1)
+	go func() {
+		cmd := exec.Command(b.path, flags...) //nolint:gosec
+		done := make(chan error)
+
+		if err := cmd.Start(); err != nil {
+			errChan <- err
+			return
+		}
+
+		go func() { done <- cmd.Wait() }()
+
+		select {
+		case <-b.interruptCh:
+			if err := cmd.Process.Signal(os.Interrupt); err != nil {
```

Comment:
One thing I'm unable to reason about with certainty looking at the code, if `swiftDialogMDMMigrator.renderMigration` is running and blocked in the `select` and the `interrupt` is triggered, that would kiil the process but would never send anything back to the channels. There's probably something missing for this case - the easiest might be to send on the errChan?

If `Process.Kill` sends an error to the result of `cmd.Wait()`, then this whole `select case` could be in its own goroutine so that killing the process triggers the send to `done` which triggers the `case err := <-done`. But that's more changes and the result is likely the same, so sending on errChan is probably easiest.

Reply:
I think you're right, I was mistakenly relying on `Process.Kill` sending a signal that would be later on caught by the `done` channel 🤦 

I gave it a shot to your idea of moving the select to a separate goroutine, just because sending to the error channel would require other changes as well.